apps/prairielearn/src/sync/sharing.sql

```diff
@@ -17,3 +18,38 @@ WHERE
         question_id = q.id
     )
   );
+
+-- BLOCK select_sharing_sets
+SELECT
+  ss.name
+FROM
+  sharing_sets AS ss
+WHERE
+  ss.course_id = $course_id;
+
+-- BLOCK select_question_sharing_sets
+WITH
+  sharing_sets_agg AS (
+    SELECT
+      ssq.question_id,
+      jsonb_agg(ss.name) AS sharing_sets
+    FROM
+      sharing_set_questions AS ssq
+      JOIN sharing_sets AS ss ON ss.id = ssq.sharing_set_id
+    GROUP BY
+      ssq.question_id
```

Comment:
This aggregates across every single question in the database, right? I can't imagine that's at all what we want? You probably want to filter on `sharing_sets.course_id` here.

Reply:
Well, it's filtering by course id after the join.

I'm not 100% sure the query planner will optimize based on that though, so I can add `where course_id =` to this part as well.

Reply:
There's actually a pretty substantial difference between the two. Here's the query plan/analysis with the code as currently written:

```
 Sort  (cost=2855.69..2855.70 rows=4 width=71) (actual time=8.070..8.074 rows=1 loops=1)
   Sort Key: q.qid
   Sort Method: quicksort  Memory: 25kB
   ->  Nested Loop  (cost=11.25..2855.65 rows=4 width=71) (actual time=6.228..8.047 rows=1 loops=1)
         ->  HashAggregate  (cost=10.83..15.10 rows=339 width=40) (actual time=0.637..0.937 rows=439 loops=1)
               Group Key: ssq.question_id
               Filter: (jsonb_agg(ss.name) IS NOT NULL)
               Batches: 1  Memory Usage: 285kB
               ->  Hash Join  (cost=1.61..9.11 rows=345 width=40) (actual time=0.053..0.225 rows=444 loops=1)
                     Hash Cond: (ssq.sharing_set_id = ss.id)
                     ->  Seq Scan on sharing_set_questions ssq  (cost=0.00..6.45 rows=345 width=16) (actual time=0.011..0.056 rows=444 loops=1)
                     ->  Hash  (cost=1.27..1.27 rows=27 width=40) (actual time=0.021..0.023 rows=28 loops=1)
                           Buckets: 1024  Batches: 1  Memory Usage: 10kB
                           ->  Seq Scan on sharing_sets ss  (cost=0.00..1.27 rows=27 width=40) (actual time=0.007..0.011 rows=28 loops=1)
         ->  Index Scan using questions_pkey on questions q  (cost=0.42..8.37 rows=1 width=39) (actual time=0.015..0.015 rows=0 loops=439)
               Index Cond: (id = ssq.question_id)
               Filter: ((deleted_at IS NULL) AND (course_id = 114))
               Rows Removed by Filter: 1
 Planning Time: 1.379 ms
 Execution Time: 8.411 ms
(20 rows)
```

And here it is with a `course_id` filter in the CTE:

```
 Sort  (cost=117.30..117.31 rows=1 width=71) (actual time=0.091..0.092 rows=1 loops=1)
   Sort Key: q.qid
   Sort Method: quicksort  Memory: 25kB
   ->  Nested Loop  (cost=7.69..117.29 rows=1 width=71) (actual time=0.083..0.084 rows=1 loops=1)
         ->  HashAggregate  (cost=7.27..7.43 rows=13 width=40) (actual time=0.055..0.056 rows=1 loops=1)
               Group Key: ssq.question_id
               Filter: (jsonb_agg(ss.name) IS NOT NULL)
               Batches: 1  Memory Usage: 24kB
               ->  Nested Loop  (cost=0.27..7.20 rows=13 width=40) (actual time=0.038..0.042 rows=1 loops=1)
                     ->  Seq Scan on sharing_sets ss  (cost=0.00..1.34 rows=1 width=40) (actual time=0.011..0.013 rows=1 loops=1)
                           Filter: (course_id = 114)
                           Rows Removed by Filter: 27
                     ->  Index Only Scan using sharing_set_questions_sharing_set_id_question_id_key on sharing_set_questions ssq  (cost=0.27..5.29 rows=58 width=16) (actual time=0.025..0.026 rows=1 loops=1)
                           Index Cond: (sharing_set_id = ss.id)
                           Heap Fetches: 0
         ->  Index Scan using questions_pkey on questions q  (cost=0.42..8.44 rows=1 width=39) (actual time=0.027..0.027 rows=1 loops=1)
               Index Cond: (id = ssq.question_id)
               Filter: ((deleted_at IS NULL) AND (course_id = 114))
 Planning Time: 0.389 ms
 Execution Time: 0.565 ms
(20 rows)
```

With the additional early filter, the query runs pretty much 10 times faster.

Reply:
Sounds good, fixed.